### ISSUE TITLE
Add `keep_existing_predictor` flag to `Finetuner` to allow full predictor replacement

### ISSUE THREAD

**Issue by dahvida** [emdgroup/matcha#81]
## Summary

Add a boolean `keep_existing_predictor` argument to `Finetuner` (default `True` for backward compatibility) that controls whether the pretrained model's predictor hidden layers are preserved before the new `pred_hidden_dims` MLP is attached. Currently there is no way to make the new MLP replace the pretrained predictor — it is always stacked on top of it. Making this a flag rather than a new `finetuning_strategy` keeps the behavior orthogonal to `"full"` / `"lora"`, so it composes cleanly with LoRA too.

## Current Behavior

`Finetuner.__init__` at `src/matcha/torch/models/finetuning/finetuner.py:119, 132-142` does:

```python
self.pretrain.predictor.prediction_head = None
...
self.pretrain_output_dim = self.pretrain.latent_dim
self.predictor = MLP(input_dim=self.pretrain_output_dim,
                     hidden_dims=pred_hidden_dims, ...)
```

- Only the pretrained predictor's final linear head is dropped; the hidden `layers` remain and are used at forward time via `self.pretrain.encode(batch)` (which internally calls `predictor.encode`).
- `pretrain_output_dim` is `pretrain.latent_dim` — i.e. the last hidden dim of the pretrained MLP, not the encoder's `fp_dim`.
- Consequence: the new `pred_hidden_dims` MLP is appended after the pretrained MLP for any pretrain origin that carries a predictor (`pretrain_multitask` classic-origin, nested `Finetuner`-origin). Only pure `origin_type: "pretraining"` artifacts (via `PretrainedEncoderWrapper`) naturally have nothing to strip, because `_StubPredictor` has no hidden layers.
- The current behavior is fixed regardless of `finetuning_strategy` (`"full"` or `"lora"`).

## Proposed Behavior

Add `keep_existing_predictor: bool = True` to `Finetuner.__init__` (and expose it through `BaseFinetuner.reload_pretrained`, `FinetuningRegressor`, `FinetuningClassifier` in `src/matcha/sklearn/finetuner.py`):

- `keep_existing_predictor=True` (default): behavior unchanged from today — the pretrained predictor's hidden layers stay in the forward path and the new `pred_hidden_dims` MLP is stacked on top.
- `keep_existing_predictor=False`: the pretrained predictor is discarded in full (both hidden layers and prediction head), and the new `pred_hidden_dims` MLP consumes the encoder's output directly. `pretrain_output_dim` becomes the leaf encoder's `fp_dim`.
- Composes with both `finetuning_strategy="full"` and `finetuning_strategy="lora"`: in each case the discarded predictor's parameters are excluded from the corresponding optimizer parameter groups. LoRA's encoder-freezing and layer injection are unaffected.
- For nested `Finetuner` pretrains (e.g. `pretrain_encoder` → `pretrain_multitask` → this finetune), `keep_existing_predictor=False` strips the intermediate wrapper's predictor as well, so only the leaf encoder feeds the new head.
- For pure `origin_type: "pretraining"` pretrains (encoder-only), `keep_existing_predictor` is a no-op — the two settings produce identical modules.

## Acceptance Criteria

- `Finetuner(keep_existing_predictor=False, pred_hidden_dims=[256, 256], ...)` builds a model whose forward path is `encoder → new [256, 256] MLP → new prediction head`, with no hidden layers from the pretrained predictor in between, for all three pretrain origins: classic (`pretrain_multitask` with GIN), pretraining (`pretrain_encoder`), and nested finetuner (`pretrain_encoder` → `pretrain_multitask` as `FinetuningRegressor`).
- Under `keep_existing_predictor=False`, `self.pretrain_output_dim` equals the leaf encoder's `fp_dim`, regardless of the pretrained predictor's `hidden_dims`.
- Under `keep_existing_predictor=False`, no parameter of the pretrained predictor appears in any optimizer parameter group.
- `keep_existing_predictor=False` works with both `finetuning_strategy="full"` and `finetuning_strategy="lora"`, including for nested-finetuner pretrains.
- `keep_existing_predictor=True` (default) reproduces existing behavior bit-for-bit for the same inputs and seed.
- For pure `"pretraining"`-origin artifacts, `keep_existing_predictor=True` and `keep_existing_predictor=False` produce identical modules and identical predictions.
- Checkpoint save + reload (via `autoload` and via the sklearn `FinetuningRegressor` / `FinetuningClassifier` paths) round-trips the `keep_existing_predictor` value and yields identical predictions.
- The sklearn wrappers (`FinetuningRegressor`, `FinetuningClassifier`) accept and forward `keep_existing_predictor`.
- Test coverage for both flag values, both strategies (`"full"` and `"lora"`), and all three pretrain origins, mirroring existing finetuner tests.

## Context

Behavior discovered while tracing what happens when `pred_hidden_dims=[256, 256]` is passed to a second finetuning stage over an already-multitask-pretrained model — the new MLP was expected to replace the multitask MLP but is actually stacked on top of it.

## Technical Notes

- Main file to change: `src/matcha/torch/models/finetuning/finetuner.py`. Between `_load_from_pretrained_path(...)` and the `self.predictor = MLP(...)` construction (roughly finetuner.py:117-142), branch on `keep_existing_predictor` — when `False`, replace the pretrained predictor (across nested `Finetuner`s if any) with a passthrough / `nn.Identity` and set `self.pretrain_output_dim = _resolve_leaf_encoder(self.pretrain).fp_dim`. `_resolve_leaf_encoder` (finetuner.py:28-32) already handles the nested case.
- Consider making the discarded-predictor branch also short-circuit `self.pretrain.encode` back to the encoder call — either by nullifying the intermediate predictors so `predictor.encode` degenerates to identity, or by overriding `forward` / `encode` when the flag is off. The former keeps the module tree simpler and interacts more predictably with checkpointing.
- Both `_setup_full` (finetuner.py:386-466) and `_setup_lora` (finetuner.py:346-384) need to skip the pretrained predictor's parameters under `keep_existing_predictor=False`. In `_setup_full` this means dropping the predictor parameter group at finetuner.py:410-415. In `_setup_lora` this affects the \"freeze everything, apply LoRA to encoder linears\" path only if the discarded predictor's params were previously counted — they were frozen anyway, so the practical change is mostly about not leaving dead modules attached.
- Sklearn side: add `keep_existing_predictor` to `BaseFinetuner.reload_pretrained`, `FinetuningRegressor.__init__`, `FinetuningClassifier.__init__` in `src/matcha/sklearn/finetuner.py` and thread through to the `Finetuner(...)` construction sites (finetuner.py:236-256 and 284-304). Update docstrings.
- Persist `keep_existing_predictor` via `save_hyperparameters()` (already covered by `HyperparametersMixin` if it's a plain `__init__` arg) so the checkpoint reload path in `_build_skeleton_from_config` / `_build_pretrain_config` handles it without extra plumbing.
- Update the docstring in finetuner.py:37-77 to document the new flag alongside the strategy descriptions.

**Comment by dahvida** [comment]
## Implementation Plan — `keep_existing_predictor` flag on `Finetuner`

**Architecture:** Pragmatic balance. Introduce a small `PassthroughPredictor` module colocated with `pretrained_encoder_wrapper.py`, one recursive strip helper next to `_resolve_leaf_encoder`, and thread the flag through the schema and sklearn wrappers. Existing tests are extended in place; no bespoke test file is added.

**Key design points:**

- `PassthroughPredictor(nn.Module)`: `.prediction_head = None`, `.encode(self, mol_features) -> mol_features`, `.forward(self, mol_features) -> mol_features`. The `mol_features` param name is required by `BaseClassicModel._unpack_batch_and_call`'s signature-based dispatch.
- Recursive helper walks any nested `Finetuner` chain and replaces every `.predictor` with a fresh `PassthroughPredictor`, so `self.pretrain.encode(batch)` naturally chains through with identity behavior at each level.
- Under `keep_existing_predictor=False`, `self.pretrain_output_dim = _resolve_leaf_encoder(self.pretrain).fp_dim` (helper already exists at `finetuner.py:28`).
- Default `True` preserves existing behavior bit-for-bit — all existing tests continue to pass unchanged.
- `save_hyperparameters()` at `finetuner.py:103` and the nested-hparams filter at `finetuner.py:334-338` already handle checkpoint round-trip; **no change to `_build_pretrain_config` is needed**.
- Under `full` strategy, `_setup_full`'s guard at `finetuner.py:410-415` already skips the predictor param group when `predictor_params` is empty — a parameter-less `PassthroughPredictor` is naturally skipped without further branching. Under `lora`, only the leaf encoder is targeted via `_resolve_leaf_encoder` — no change needed.
- Scope is main `Finetuner` only; `ChempropFinetuner` is out of scope.

## Staged breakdown

### Stage 1 — Torch layer: schema, `PassthroughPredictor`, `Finetuner` branch, torch tests

**Create:**
- `src/matcha/torch/models/finetuning/passthrough_predictor.py` — `PassthroughPredictor(nn.Module)` with `.prediction_head=None`, `.encode(self, mol_features)`, `.forward(self, mol_features)`. Docstring parallel to `_StubPredictor` explaining its role in the Finetuner-compat adapter surface.

**Modify:**
- `src/matcha/utils/schemas/generic_models.py` — `FinetunerMixin` (line 247-267): add `keep_existing_predictor: bool = True` with description.
- `src/matcha/torch/models/finetuning/finetuner.py`:
  - Add `keep_existing_predictor: bool = True` to `Finetuner.__init__` (place after `finetuning_strategy`).
  - Add module-level helper `_strip_pretrained_predictor_recursive(pretrain: nn.Module) -> None` next to `_resolve_leaf_encoder` (line 28-32). Walks the nested `Finetuner` chain and installs a fresh `PassthroughPredictor` at every level and on the leaf.
  - Replace the `self.pretrain.predictor.prediction_head = None` + `self.pretrain_output_dim = self.pretrain.latent_dim` block (lines 119, 132) with a branch on `keep_existing_predictor`. When `False`, call the strip helper and use `_resolve_leaf_encoder(self.pretrain).fp_dim`.
  - Expand the class docstring (lines 37-77) to document the flag as an axis orthogonal to `finetuning_strategy`.

**Extend tests:**
- `tests/torch/models/finetuning/test_finetuner_nested_encoder.py`:
  - Add `keep_existing_predictor` to `_make_nested_finetuner`; parametrize new test classes over (origin × strategy × flag). Origins covered: classic (`pretrain_multitask` with GIN), pretraining (`pretrain_encoder`), nested finetuner (`pretrain_encoder` → `pretrain_multitask` → this finetune).
  - Assertions:
    - Under `False`: every `.predictor` in the chain is a `PassthroughPredictor`; `pretrain_output_dim == _resolve_leaf_encoder(...).fp_dim`; forward pass returns correct output shape; no pretrained-predictor param appears in `_setup_full`'s `param_groups`.
    - Under `True`: existing behavior preserved (structure intact, prediction_head=None on leaf only).
    - Pure `pretraining` origin: both flag values produce numerically-equal outputs on the same input (AC: identical modules and identical predictions).
    - `Finetuner.load_from_checkpoint` on a saved-with-`False` finetuner reproduces the `PassthroughPredictor` chain and yields identical predictions.

**Testing:** unit + integration via the in-memory `_SELF_CONTAINED_SENTINEL` skeleton-config path. `uv run pytest -k 'not gpu' tests/torch/models/finetuning/`.

**Dependencies:** none.

---

### Stage 2 — Sklearn layer thread-through + one end-to-end round-trip test

**Modify:**
- `src/matcha/sklearn/finetuner.py`:
  - `BaseFinetuner.reload_pretrained` (line 73-105): add `keep_existing_predictor: bool = True` to signature and docstring; forward to both `Finetuner(...)` call sites — pretraining branch (line 236-256) and classic branch (line 284-304). Chemprop branch (line 267-281) intentionally skipped.
  - `FinetuningRegressor.__init__` (line 374-406): add param; forward at line 408.
  - `FinetuningClassifier.__init__` (line 453-485): add param; forward at line 493.

**Extend tests:**
- `tests/sklearn_misc/test_finetuner.py`:
  - Add one end-to-end test: build a classic pretrained regressor via the existing factory, `save_model`, construct `FinetuningRegressor(..., keep_existing_predictor=False)`, `fit` on the mini dataset, `save_model`, `autoload`, `predict`, and assert predictions are stable across the save/autoload cycle. Assert `isinstance(loaded._model.pretrain.predictor, PassthroughPredictor)` on the reloaded model.
  - Narrow by design: broader (origin × strategy × flag) matrix is covered at the torch layer in Stage 1.

**Testing:** integration end-to-end via `uv run pytest -k 'not gpu' tests/sklearn_misc/test_finetuner.py`.

**Dependencies:** Stage 1 (torch layer must be in place).

---

### Stage 3 — Documentation

**Modify:**
- `docs/source/` — locate the finetuning-related recipe or getting-started file that references the finetuner and add a short subsection describing `keep_existing_predictor`, when to set it to `False` (e.g., second-stage finetuning where the caller wants to fully replace the previous MLP rather than stack on top of it), and its orthogonality with `finetuning_strategy`. Note that for pure-encoder pretraining origins the flag is a no-op.

**Testing:** none (docs only). Ensure any docs-build CI step passes.

**Dependencies:** Stages 1-2 complete (so docstring/example references match the working API).

---

## Layer coverage check

| Layer | Addressed by |
|---|---|
| Pydantic schema (`utils/schemas/generic_models.py`) | Stage 1 |
| Torch model (`torch/models/finetuning/finetuner.py`, `passthrough_predictor.py`) | Stage 1 |
| Torch tests (`tests/torch/models/finetuning/`) | Stage 1 |
| Sklearn wrappers (`sklearn/finetuner.py`) | Stage 2 |
| Sklearn tests (`tests/sklearn_misc/`) | Stage 2 |
| Documentation (`docs/source/`) | Stage 3 |
| CLI | not affected — schema-driven, auto-accepts new field |
| Managers / registries | not affected |

## Decision Log

**Clarifying questions (Step 3):**

- *How should the pretrained predictor be stripped when `keep_existing_predictor=False`?* — Use a `PassthroughPredictor` `nn.Module` (`.prediction_head=None`, `.encode(x)=x`, `.forward(x)=x`) installed in place of `self.pretrain.predictor` and recursively across nested `Finetuner`s. Rationale: both `BaseClassicModel.encode` (base_classic_model.py:171) and `Finetuner.encode` (finetuner.py:494) dispatch through `predictor.encode`; using `nn.Identity` would break them (no `.encode` method). This choice also matches the issue's technical-notes preference for keeping the module tree simple and predictable for checkpointing.
- *Should the flag also be added to `ChempropFinetuner`?* — No, main `Finetuner` only. Rationale: `ChempropFinetuner` follows a completely different loading path (delegates into chemprop internals) and the issue's ACs are worded around the three origins of the main `Finetuner`.
- *How deep should the docs update go?* — Docstrings plus a section in `docs/source/`. Rationale: the flag is user-facing and affects a common workflow (second-stage finetuning), so a short guide entry pays for itself.
- *What depth of checkpoint round-trip testing?* — Torch-layer `load_from_checkpoint` round-trip covering (origin × strategy × flag) in Stage 1, plus one sklearn end-to-end (`fit → save_model → autoload → predict`) in Stage 2. Rationale: the (origin × strategy × flag) matrix is cheap at the torch layer via the in-memory skeleton-config path; one sklearn end-to-end is enough to prove the full path works without ballooning CI runtime.

**Architecture choice (Step 4): Pragmatic balance.**

Chosen because:
1. `PassthroughPredictor` deserves its own file next to `pretrained_encoder_wrapper.py::_StubPredictor` — they are the same architectural concept (adapter surface the `Finetuner` expects on `.pretrain.predictor`).
2. One extracted helper (`_strip_pretrained_predictor_recursive`) is enough; a second `_configure_pretrain_predictor` wrapper (proposed in the clean-architecture variant) wraps a 3-line branch without a clear payoff.
3. Extending the existing `tests/torch/models/finetuning/test_finetuner_nested_encoder.py` reuses the `_make_nested_finetuner`, `GIN_PRETRAIN_PARAMS`, and `GIN_PRETRAINING_PARAMS` scaffolding, avoiding duplicate test infrastructure.
4. 3 CLI-session-sized stages align naturally with the code layers (torch, sklearn, docs).

Alternatives considered:
- **Minimal changes** (inline `PassthroughPredictor` in `finetuner.py`): rejected — violates the local convention of placing adapter surfaces in their own modules (`_StubPredictor` lives in `pretrained_encoder_wrapper.py`).
- **Clean architecture** (separate module + two extracted helpers + new dedicated test file): rejected — the second helper (`_configure_pretrain_predictor`) is unnecessary indirection, and a new test file duplicates existing scaffolding.

---

*This comment will guide staged implementation via `/mach10:issue-implement 81 <stage>`.*

**Comment by dahvida** [comment]
### Progress

Added `keep_existing_predictor: bool = True` to `Finetuner` (and mirrored on `FinetunerMixin` schema).

- When `False`, a new `PassthroughPredictor` replaces every `.predictor` in the pretrain chain — including nested `Finetuner` wrappers and the leaf model — so the new `pred_hidden_dims` head consumes the leaf encoder's `fp_dim` output directly.
- When `True` (default), historic behavior is preserved: the pretrained predictor's hidden layers stay wired up and only its final `prediction_head` is dropped.
- For pure `origin_type="pretraining"` artifacts the flag is a no-op (there is no pretrained predictor to strip) — verified with a `torch.allclose` parity test.

Tests added in `tests/torch/models/finetuning/test_finetuner_nested_encoder.py`:
- flat classic + pretraining origins under both `full` / `lora`,
- nested finetuner chain,
- optimizer group hygiene (no stripped params leak in),
- forward shapes,
- skeleton round-trip predictions identical before/after reload.

Commit: 3778fbe
